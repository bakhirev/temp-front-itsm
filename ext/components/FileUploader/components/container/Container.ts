import styled from 'styled-components';

export const Container = styled.div<{ mobile?: boolean }>`
  width: ${({ mobile }) => (mobile ? '100%' : '384px')};
`;
