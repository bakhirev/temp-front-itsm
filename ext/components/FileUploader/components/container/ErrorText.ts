import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../../../common';
import { Body2Short } from '../../../Typography';

export const ErrorText = styled(Body2Short)`
  ${({ theme }) => css`
    color: ${theme.color.error[60]};
    margin-top: 8px;
  `}
`;

ErrorText.defaultProps = {
  theme: DEFAULT_THEME,
};
