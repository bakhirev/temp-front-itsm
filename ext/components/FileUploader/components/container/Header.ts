import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../../../common';
import { Subtitle3, Body2Short } from '../../../Typography';

export interface IHeaderTitleProps {
  disabled?: boolean;
}

export const HeaderTitle = styled(Subtitle3)<IHeaderTitleProps>`
  ${({ theme, disabled }) => css`
    ${disabled ? `color: ${theme.color.neutral[30]};` : ''}
    margin-bottom: 2px;
  `}
`;

HeaderTitle.defaultProps = {
  theme: DEFAULT_THEME,
};

export interface IHeaderSubtitleProps {
  disabled?: boolean;
}

export const HeaderSubtitle = styled(Body2Short)<IHeaderSubtitleProps>`
  ${({ theme, disabled }) => css`
    color: ${disabled ? theme.color.neutral[30] : theme.color.neutral[50]};
    margin-bottom: 16px;
  `}
`;

HeaderSubtitle.defaultProps = {
  theme: DEFAULT_THEME,
};
