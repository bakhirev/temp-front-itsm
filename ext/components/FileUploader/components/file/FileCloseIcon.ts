import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../../../common';

export const FileCloseIcon = styled.div`
  ${({ theme }) => css`
    height: 20px;
    width: 20px;
    cursor: pointer;

    & svg {
      fill: ${theme.color.neutral[50]};
      height: 20px;
      width: 20px;
      transition: all 0.2s;

      &:hover {
        fill: ${theme.color.primary[60]};
      }
    }
  `}
`;

FileCloseIcon.defaultProps = {
  theme: DEFAULT_THEME,
};
