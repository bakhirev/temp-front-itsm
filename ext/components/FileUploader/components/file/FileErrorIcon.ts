import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../../../common';

export const FileErrorIcon = styled.div`
  ${({ theme }) => css`
    height: 20px;
    width: 20px;
    & svg {
      fill: ${theme.color.error[60]};
      height: 20px;
      width: 20px;
    }
  `}
`;

FileErrorIcon.defaultProps = {
  theme: DEFAULT_THEME,
};
