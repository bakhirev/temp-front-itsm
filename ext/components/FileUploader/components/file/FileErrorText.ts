import styled, { css } from 'styled-components';

import { Body2Short } from '../../../Typography';
import { DEFAULT_THEME } from '../../../common';

export const FileErrorText = styled(Body2Short)`
  ${({ theme }) => css`
    color: ${theme.color.error[60]};
    margin-top: 8px;
  `}
`;

FileErrorText.defaultProps = {
  theme: DEFAULT_THEME,
};
