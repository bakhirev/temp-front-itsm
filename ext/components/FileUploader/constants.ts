export const UPLOAD_TEXT = 'Перетащите файлы в эту область или нажмите для загрузки';
export const FILE_MAX_SIZE_ERROR = 'Размер файла превышает разрешенный размер';
export const FILE_MIN_SIZE_ERROR = 'Размер файла меньше минимального размера';
export const FILE_EXTENSION_ERROR = 'Файл не соответствует допустимым расширениям';
export const GET_FILE_COUNT_EXCEED_ERROR = `Несколько файлов не было загружено, так как превышен лимит количества загружаемых файлов`;
export const GET_FILE_TOTAL_SIZE_ERROR = (megabytes: number) =>
  `Суммарно размер файлов не должен превышать ${megabytes} Mb`;
export const DEFAULT_MAX_TOTAL_SIZE = 1024 * 1024 * 100;
export const UPLOAD_TEXT_MOBILE = 'Нажмите для загрузки';
