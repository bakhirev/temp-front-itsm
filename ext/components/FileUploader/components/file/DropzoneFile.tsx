import React, { FC, useRef, useState, useEffect, RefObject } from 'react';

import { createTooltipHOC } from '../../../TooltipHOC';
import { ReactComponent as ErrorSolid } from '../../../Icons/service/ErrorSolid.svg';
import { ReactComponent as CloseOutline } from '../../../Icons/service/CloseOutline.svg';
import { ReactComponent as CheckSolid } from '../../../Icons/service/CheckSolid.svg';

import { FileText } from './FileText';
import { FileWrapper } from './FileWrapper';
import { FileContent } from './FileContent';
import { FileTextWrapper } from './FileTextWrapper';
import { FileIconsWrapper } from './FileIconsWrapper';
import { FileErrorIcon } from './FileErrorIcon';
import { FileErrorText } from './FileErrorText';
import { FileCloseIcon } from './FileCloseIcon';
import { FileSuccessIcon } from './FileSuccessIcon';
import { WrapperCloseIcon } from './WrapperCloseIcont';

export interface IDropzoneFile {
  /** Название файла */
  name: string;
  /** Файл */
  data?: File;
  /** Файл в состоянии ошибки */
  error?: string;
  /** Файл в состоянии прогресса */
  progress?: number;
  /** Файл в состоянии загрузки */
  loading?: boolean;
}

interface IFileProps {
  onRemoveClick: () => void;
  data: IDropzoneFile;
  waitingLoad: boolean;
  tooltipContainer?: Element | null;
}

export const DropzoneFile: FC<IFileProps> = ({
  data: { name, error, progress },
  onRemoveClick,
  waitingLoad,
  tooltipContainer,
}) => {
  const [overflow, setOverflow] = useState(false);
  const contentRef: RefObject<HTMLDivElement> = useRef(null);

  const detectOverflow = (e: any) => e.offsetWidth < e.scrollWidth;

  useEffect(() => {
    const contentCurrent = contentRef.current;
    if (contentCurrent && detectOverflow(contentCurrent) !== overflow) {
      setOverflow(detectOverflow(contentCurrent));
    }
  }, []);

  const Text = overflow ? createTooltipHOC(FileText) : FileText;

  const shouldRenderSuccessIcon = !waitingLoad && !error;
  const shouldRenderErrorIcon = error;
  const shouldRenderCloseIcon = !waitingLoad || error;
  return (
    <FileWrapper>
      <FileContent progress={progress}>
        <FileTextWrapper>
          <Text
            ref={contentRef}
            tooltip={name}
            isWaitingLoad={waitingLoad}
            container={tooltipContainer}
          >
            {name}
          </Text>
        </FileTextWrapper>
        <FileIconsWrapper>
          {shouldRenderErrorIcon && (
            <FileErrorIcon>
              <ErrorSolid />
            </FileErrorIcon>
          )}
          {shouldRenderSuccessIcon && (
            <FileSuccessIcon>
              <CheckSolid />
            </FileSuccessIcon>
          )}
          {shouldRenderCloseIcon && (
            <WrapperCloseIcon>
              <FileCloseIcon onClick={onRemoveClick}>
                <CloseOutline />
              </FileCloseIcon>
            </WrapperCloseIcon>
          )}
        </FileIconsWrapper>
      </FileContent>
      {error && <FileErrorText>{error}</FileErrorText>}
    </FileWrapper>
  );
};
