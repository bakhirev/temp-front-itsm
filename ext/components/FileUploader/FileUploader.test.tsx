import React from 'react';
import { shallow, mount } from 'enzyme';
import { render, fireEvent } from '@testing-library/react';

import { FILE_MAX_SIZE_ERROR, FILE_EXTENSION_ERROR } from './constants';
import { FileCloseIcon } from './components/file/FileCloseIcon';
import { FileErrorText } from './components/file/FileErrorText';
import { FileErrorIcon } from './components/file/FileErrorIcon';
import { ErrorText } from './components/container/ErrorText';

import { FileUploader } from './';

const DEFAULT_MAX_TOTAL_SIZE = 20 * 1024 * 1024;

jest.mock('../common', () => ({
  DEFAULT_THEME: {
    borderRadius: '',
    color: {
      error: {
        60: '',
      },
      neutral: {
        10: '',
        30: '',
      },
      primary: {
        60: '',
      },
      success: {
        20: '',
      },
      opacity: {
        blackHover: '',
      },
    },
  },
  Z_INDEXES: {
    TOOLTIP: 777,
  },
}));

describe('FileUploader', () => {
  const testFile = new File([JSON.stringify({ ping: true })], 'ping.json', {
    type: 'application/json',
  });

  beforeEach(() => {
    // Remove it when DropZone console error will be fixed
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    jest.spyOn(console, 'error').mockImplementation(() => {});
  });

  it('should correctly render', () => {
    const wrapper = shallow(<FileUploader maxTotalSize={DEFAULT_MAX_TOTAL_SIZE} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should correctly render if disabled', () => {
    const wrapper = shallow(<FileUploader disabled maxTotalSize={DEFAULT_MAX_TOTAL_SIZE} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with mobile prop', () => {
    const wrapper = shallow(<FileUploader maxTotalSize={DEFAULT_MAX_TOTAL_SIZE} mobile />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onChange on file drops', async () => {
    const mockChange = jest.fn();
    const data = mockData([testFile]);
    const ui = <FileUploader onChange={mockChange} maxTotalSize={DEFAULT_MAX_TOTAL_SIZE} />;
    await initDrop(ui, data);
    expect(mockChange).toHaveBeenCalled();
  });

  it('should not call onChange on file drops, with mobile prop', async () => {
    const mockChange = jest.fn();
    const data = mockData([testFile]);
    const ui = <FileUploader onChange={mockChange} maxTotalSize={DEFAULT_MAX_TOTAL_SIZE} mobile />;
    await initDrop(ui, data);
    expect(mockChange).toBeCalledTimes(0);
  });

  it('should call onChange with limited files count', async () => {
    const mockChange = jest.fn();
    const secondFile = new File([JSON.stringify({ ping: true })], 'pinger.json', {
      type: 'application/json',
    });
    const data = mockData([testFile, secondFile]);
    const ui = (
      <FileUploader maxFiles={1} onChange={mockChange} maxTotalSize={DEFAULT_MAX_TOTAL_SIZE} />
    );
    await initDrop(ui, data);
    expect(mockChange).toHaveBeenCalledWith([{ data: testFile, name: testFile.name }]);
  });

  it('should call onChange with error because of max size', async () => {
    const mockChange = jest.fn();
    const data = mockData([testFile]);
    const ui = <FileUploader maxSize={1} onChange={mockChange} maxTotalSize={1} />;
    await initDrop(ui, data);
    expect(mockChange).toHaveBeenCalledWith([
      { data: testFile, error: FILE_MAX_SIZE_ERROR, name: testFile.name },
    ]);
  });

  it('should call onRemoveClick', () => {
    const mockRemove = jest.fn();
    const testFile = { name: 'testFile.jpg' };
    const wrapper = mount(
      <FileUploader
        files={[testFile]}
        onRemoveClick={mockRemove}
        maxTotalSize={DEFAULT_MAX_TOTAL_SIZE}
      />
    );
    wrapper.find(FileCloseIcon).simulate('click');
    expect(mockRemove).toHaveBeenCalledWith(testFile, 0);
  });

  it('should render file error', () => {
    const testFile = { name: 'testFile.jpg', error: 'some error text' };
    const wrapper = mount(
      <FileUploader files={[testFile]} maxTotalSize={DEFAULT_MAX_TOTAL_SIZE} />
    );
    expect(wrapper.find(FileErrorText).text()).toEqual(testFile.error);
    expect(wrapper.find(FileErrorIcon).length).toEqual(1);
  });

  it('should render dropzone error', () => {
    const errorText = 'some error';
    const wrapper = mount(<FileUploader error={errorText} maxTotalSize={DEFAULT_MAX_TOTAL_SIZE} />);
    expect(wrapper.find(ErrorText).text()).toEqual(errorText);
  });

  it('should call onChange with error if file extension not correct', async () => {
    const mockChange = jest.fn();
    const data = mockData([testFile]);
    const ui = (
      <FileUploader
        accept={['image/png']}
        onChange={mockChange}
        maxTotalSize={DEFAULT_MAX_TOTAL_SIZE}
      />
    );
    await initDrop(ui, data);
    expect(mockChange).toHaveBeenCalledWith([
      { data: testFile, error: FILE_EXTENSION_ERROR, name: testFile.name },
    ]);
  });
});

async function initDrop(ui: any, data: any) {
  const { container } = render(ui);
  const dropzone = (container.querySelector('input') as any).parentElement;
  dispatchEvt(dropzone, 'drop', data);
  await flushPromises(ui, container);
}

function flushPromises(ui: any, container: any) {
  return new Promise((resolve) =>
    setImmediate(() => {
      render(ui, { container });
      resolve(container);
    })
  );
}

function dispatchEvt(node: any, type: any, data: any) {
  const event = new Event(type, { bubbles: true });
  Object.assign(event, data);
  fireEvent(node, event);
}

function mockData(files: any) {
  return {
    dataTransfer: {
      files,
      items: files.map((file: any) => ({
        kind: 'file',
        type: file.type,
        getAsFile: () => file,
      })),
      types: ['Files'],
    },
  };
}
