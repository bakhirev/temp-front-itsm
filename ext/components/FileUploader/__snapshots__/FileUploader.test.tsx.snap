// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUploader should correctly render 1`] = `
<styled.div>
  <Component
    files={Array []}
    maxSize={20971520}
    maxTotalSize={20971520}
    onChange={[Function]}
    onFileCountLimitExceed={[Function]}
    onFileTotalSizeLimitExceed={[Function]}
    uploadText="Перетащите файлы в эту область или нажмите для загрузки"
  />
</styled.div>
`;

exports[`FileUploader should correctly render if disabled 1`] = `
<styled.div>
  <Component
    disabled={true}
    files={Array []}
    maxSize={20971520}
    maxTotalSize={20971520}
    onChange={[Function]}
    onFileCountLimitExceed={[Function]}
    onFileTotalSizeLimitExceed={[Function]}
    uploadText="Перетащите файлы в эту область или нажмите для загрузки"
  />
</styled.div>
`;

exports[`FileUploader should render with mobile prop 1`] = `
<styled.div
  mobile={true}
>
  <Component
    files={Array []}
    maxSize={20971520}
    maxTotalSize={20971520}
    mobile={true}
    onChange={[Function]}
    onFileCountLimitExceed={[Function]}
    onFileTotalSizeLimitExceed={[Function]}
    uploadText="Нажмите для загрузки"
  />
</styled.div>
`;
